challenge_03/base_solution.py

```diff
@@ -0,0 +1,97 @@
+import typing
+from typing import Tuple, Union
+
+import torch
+import torch.nn as nn
+
+
+class GraphProd2Vec(nn.Module):
+    """
+    p-companion - main stage section 4.1
+    https://assets.amazon.science/d5/16/3f7809974a899a11bacdadefdf24/p-companion-a-principled-framework-for-diversified-complementary-product-recommendation.pdf
+    """
+
+    def __init__(self, in_embedding_len: int, out_embedding_len: int):
+        super().__init__()
+        self.in_embedding_len = in_embedding_len
+        self.out_embedding_len = out_embedding_len
+        self.encoder = self._FFN_encoder()
+
+    def _FFN_encoder(self) -> nn.Module:
+        w_1 = nn.Linear(self.in_embedding_len, self.in_embedding_len)
+        bn_1 = nn.LayerNorm(self.in_embedding_len)
+        tanh_1 = nn.Tanh()
+        w_2 = nn.Linear(self.in_embedding_len, self.in_embedding_len)
+        bn_2 = nn.LayerNorm(self.in_embedding_len)
+        tanh_2 = nn.Tanh()
+        w_3 = nn.Linear(self.in_embedding_len, self.out_embedding_len)
+        bn_3 = nn.LayerNorm(self.out_embedding_len)
+        return nn.Sequential(w_1, bn_1, tanh_1, w_2, bn_2, tanh_2, w_3, bn_3)
```

Comment:
už předtím jsi na to měl nějakou funkci ne?

Reply:
frajer! tiež ma to napadlo :D len som chcel nechať tuto ulohu ako uzavretu